jablib/src/main/java/org/jabref/logic/util/LocationDetector.java

```diff
@@ -0,0 +1,192 @@
+package org.jabref.logic.util;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import org.jabref.logic.JabRefException;
+import org.jabref.logic.journals.ltwa.NormalizeUtils;
+
+import org.jspecify.annotations.NonNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * A singleton utility class for detecting and extracting city and country names from given text.
+ *
+ * <p>
+ * The location data is loaded from the resource file {@code /util/countries_cities1000.txt}
+ * and is processed to optimize matching performance by tracking minimum and maximum location
+ * name lengths.
+ * </p>
+ *
+ * <p>
+ * The data is sourced from <a href="https://www.geonames.org/export/">GeoNames</a> and lists all
+ * cities with population > 1000. Country names are manually appended at the end to get the consolidated
+ * list of locations.
+ * </p>
+ *
+ * <p>Usage example:</p>
+ * <pre>{@code
+ * LocationDetector detector = LocationDetector.getInstance();
+ * Set<String> locations = detector.extractLocations("I traveled from New York to Paris");
+ * // Returns: {"new york", "paris"}
+ * }</pre>
+ */
+public class LocationDetector {
+    private static final Logger LOGGER = LoggerFactory.getLogger(LocationDetector.class);
+    private static final String LOCATIONS_FILE = "/util/countries_cities1000.txt";
+    private static final Pattern TOKEN_PATTERN = Pattern.compile("[\\p{L}\\p{N}.'-]+");
+
+    private final Set<String> locations = new HashSet<>();
+    private int maxLocationLength = 0;
+    private int minLocationLength = Integer.MAX_VALUE;
+
+    private LocationDetector() throws JabRefException {
```

Comment:
This whole class should be loading the data from an mv file like for example we do it with nournal abbrvatiosn or other data